dockerfiles/ruby/solargraph.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+// lazyObjectStream wraps a jsonrpc2.ObjectStream to only start the connection
+// once we have received an initialize request.
+type lazyObjectStream struct {
+	Connect func(rootURI string) (jsonrpc2.ObjectStream, error)
+
+	once   *sync.Once    // protects writes to stream, err and closing ready.
+	ready  chan struct{} // protects reads to stream and err.
+	stream jsonrpc2.ObjectStream
+	err    error
+}
+
+func (t *lazyObjectStream) WriteObject(v interface{}) error {
+	t.once.Do(func() {
+		defer close(t.ready)
+
+		// minimal json to get at rootUri
+		var msg struct {
+			Method string `json:"method"`
+			Params struct {
+				RootURI string `json:"rootUri"`
+			} `json:"params"`
+		}
+		b, err := json.Marshal(v)
+		if err != nil {
+			t.err = err
+			return
+		}
+		if err = json.Unmarshal(b, &msg); err != nil {
+			t.err = err
+			return
+		}
+
+		if msg.Method != "initialize" {
+			t.err = fmt.Errorf("expected first message to be initialize, got %s", msg.Method)
+			return
+		}
+
+		t.stream, t.err = t.Connect(msg.Params.RootURI)
+	})
+	if t.err != nil {
+		return t.err
+	}
+	return t.stream.WriteObject(v)
+}
+
+func (t *lazyObjectStream) ReadObject(v interface{}) error {
+	// we wait for the initialize request (via WriteObject) to happen first.
+	<-t.ready
+	if t.err != nil {
+		return t.err
+	}
+	return t.stream.ReadObject(v)
+}
+
+func (t *lazyObjectStream) Close() error {
+	t.once.Do(func() {
```

Comment:
Do you need multiple instances of `once`? Won't the different `Do` calls interfere?

Reply:
see the doc for this once. It protects writes on stream, err and ready. If we close before we have opened the stream we need to prevent the open from happening.